### PR TITLE
Consolidate Tools into a single Binary

### PR DESCRIPTION
**Pull request by charlespierce**
Info
-----
* Currently, we are building separate binary packages for each of the core tools that we support (`node`, `npm`, `npx`, `yarn`), as well as for `launchbin` and `launchscript`.
* Each of these binaries includes most of the core of the Notion functionality, so they are fairly redundant with each other.
* As a result of including so many redundant binaries, the installer is quite large (~65 MB on Unix currently)
* We can save installer space as well as make Notion more adaptable to new tools going forward by consolidating these into a single executable.

Changes
-----
* Updated the `Tool` trait and implementations to support receiving the command-line arguments as inputs, instead of getting them internally.
* Updated `Tool::exec` to return a result, instead of exiting.
* Added `execute_tool` function that parses the command-line arguments to determine which tool should be executed, then passes control over to that tool.
* Removed redundant binaries in favor of a single `shim` binary.
* Updated Unix installer to create symlinks to the `shim` binary named `node`, `npm`, `npx`, and `yarn`.
* Updated Unix installer to handle the case of existing shims, by updating the symlink to point to `shim` instead of `launchbin`.
* Refactored error display to reduce the duplication of `if err.is_user_friendly()` blocks.

Notes
-----
* Windows installer work wasn't tackled here. If #268 is merged first then I will update this to include Windows support, if this is merged first I will update that one.
* After this update, the Unix installer is ~21.5 MB, so it's roughly one-third the size.

### CODE REVIEW

**Comment by charlespierce** [comment]
AppVeyor tests keep failing, going to have to jump onto my windows development machine to test and figure out what's going on

File: crates/notion-core/src/style.rs [hunk 1]
```diff
@@ -85,9 +85,11 @@ fn display_development_details(err: &NotionError) {
     eprintln!("{} {:?}", style("details:").yellow().bold(), err);
     eprintln!();
 
-    // For now, we require RUST_BACKTRACE for this to work.
-    // See: https://github.com/notion-cli/notion/issues/75
-    eprintln!("{:?}", err.backtrace());
+    // If `RUST_BACKTRACE` is set, then the backtrace will be included in the above output
+    // If not, we should let the user know how to see the backtrace
+    if env::var("RUST_BACKTRACE").is_err() {
+        eprintln!("Run with NOTION_DEV=1 and RUST_BACKTRACE=1 for a backtrace.");
```
Comment (dschmidt) [inline crates/notion-core/src/style.rs]: :tada:

File: Cargo.toml [hunk 2]
```diff
@@ -18,28 +18,8 @@ name = "notion"
...
-path = "src/yarn.rs"
-
-[[bin]]
-name = "launchbin"
-path = "src/launchbin.rs"
-
-[[bin]]
-name = "launchscript"
-path = "src/launchscript.rs"
+name = "shim"
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply Cargo.toml]: I would say not, though maybe @dherman would have a better idea as to why that wasn't the case already. The one argument I can see if having it named `notion` to align with the binary name makes it slightly easier when you're looking for the entry point to the `notion` binary, with any real downsides.

File: crates/notion-core/src/path/unix.rs [hunk 3]
```diff
@@ -111,12 +111,8 @@ pub fn notion_file() -> Fallible<PathBuf> {
...
 }
 
-pub fn launchbin_file() -> Fallible<PathBuf> {
-    Ok(notion_home()?.join("launchbin"))
-}
-
-pub fn launchscript_file() -> Fallible<PathBuf> {
-    Ok(notion_home()?.join("launchscript"))
+pub fn shim_executable() -> Fallible<PathBuf> {
+    Ok(notion_home()?.join("shim"))
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/notion-core/src/path/unix.rs]: I don't particularly have a preference, I think I would lean towards the `map` approach as well, but this matches the approach used almost everywhere else in the `path` module.

File: crates/notion-core/src/style.rs [hunk 4]
```diff
@@ -78,6 +81,17 @@ pub fn display_unknown_error<E: Fail>(cx: ErrorContext, err: &E) {
     }
 }
 
+fn display_development_details(err: &NotionError) {
+    eprintln!("{} {:?}", style("details:").yellow().bold(), err);
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/notion-core/src/style.rs]: My personal opinion: I tried this out, and I think that having `\n` intermixed with the implicit `\n` provided by `eprintln!` makes it confusing. I think that if we're going to explicitly include `\n`, we should go with the latter approach of dropping down to `eprint!` and including the desired number. If we stick with `eprintln!`, we should let the macro handle all newlines, including using `eprintln!()` as a blank line.

File: crates/notion-core/src/path/unix.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chriskrycho) [reply crates/notion-core/src/path/unix.rs]: Yep, and to be clear I offered it as a question for exactly that reason. "Do what the codebase does unless everyone thinks it's worth switching" is my default.

File: crates/notion-core/src/tool/mod.rs [hunk 5]
```diff
@@ -102,27 +96,32 @@ impl BinaryExecError {
     }
 }
 
+pub fn execute_tool(session: &mut Session) -> Fallible<ExitStatus> {
+    let mut args = args_os();
+    let exe = get_tool_name(&mut args)?;
+
+    match &exe.to_str() {
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/notion-core/src/tool/mod.rs]: Agreed, updated. Let me know if I could be more precise in my description, I'm still a Rust newbie in a lot of the more advanced details :+1:

File: crates/notion-core/src/tool/mod.rs [hunk 6]
```diff
@@ -134,29 +133,50 @@ pub trait Tool: Sized {
...
+}
+
+#[derive(Debug, Fail, NotionFail)]
+#[fail(display = "Tool name could not be determined")]
+#[notion_fail(code = "UnknownError")]
+struct CouldNotDetermineTool;
+
+fn get_tool_name(args: &mut ArgsOs) -> Fallible<OsString> {
+    let opt = args
+        .next()
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/notion-core/src/tool/mod.rs]: Updated, though in this case it's `nth(0)`, since we want the first element.

File: crates/notion-core/src/tool/mod.rs [hunk 7]
```diff
@@ -134,29 +133,50 @@ pub trait Tool: Sized {
...
+        .and_then(|arg0| Path::new(&arg0).file_name().map(tool_name_from_file_name));
+    if let Some(tool_name) = opt {
+        Ok(tool_name)
+    } else {
+        throw!(CouldNotDetermineTool);
+    }
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/notion-core/src/tool/mod.rs]: Agreed here, since we're already using `and_then`, it's cleaner using `ok_or`. Updated

File: crates/notion-core/src/tool/mod.rs [hunk 8]
```diff
@@ -134,29 +133,50 @@ pub trait Tool: Sized {
...
+#[cfg(unix)]
+fn tool_name_from_file_name(file_name: &OsStr) -> OsString {
+    file_name.to_os_string()
+}
+
+#[cfg(windows)]
+fn tool_name_from_file_name(file_name: &OsStr) -> OsString {
+    // On Windows PowerShell, the file name includes the .exe suffix
+    // We need to remove that to get the raw tool name
+    let mut result = OsString::new();
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/notion-core/src/tool/mod.rs]: Good catch. I went back and forth on this conversion so many times, I think I missed the easy approach :+1:

File: crates/notion-core/src/tool/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (chriskrycho) [reply crates/notion-core/src/tool/mod.rs]: Doh. I always trip myself up on 0th/1st argument.

File: crates/notion-core/src/tool/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (chriskrycho) [reply crates/notion-core/src/tool/mod.rs]: Doh. I always trip myself up on 0th/1st argument. I blame my early and formative experience with Fortran and its 1-indexed arrays. 😏

File: crates/notion-core/src/tool/mod.rs [hunk 9]
```diff
@@ -100,6 +100,10 @@ pub fn execute_tool(session: &mut Session) -> Fallible<ExitStatus> {
     let mut args = args_os();
     let exe = get_tool_name(&mut args)?;
 
+    // There is some duplication in the calls to `.exec` here.
```
Comment (chriskrycho) [inline crates/notion-core/src/tool/mod.rs]: The explanatory comment you added here is great!

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dherman) [reply Cargo.toml]: It seems nice to stick with conventions, so maybe it's nicer to use `main.rs`.

File: crates/notion-core/src/path/unix.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dherman) [reply crates/notion-core/src/path/unix.rs]: I tend to like using `?` syntax to use Rust as if it had exceptions, effectively treating `Result` as a lower layer of abstraction. (What I really wish is that [catch functions](https://internals.rust-lang.org/t/pre-rfc-catching-functions/6505) existed.) That's why I've used this style.

[the remark this answers is not in the document]
Comment (chriskrycho) [reply crates/notion-core/src/path/unix.rs]: Ah, I see! (We have differences on that point, as anyone bold and persevering enough to read through that thread knows. 😜)

As I noted above, I'm in general very happy to defer to the codebase standard, and I like `?` in general… but I admit that I find expressions with a `?` *inside* an `Ok` really hard to parse mentally. This one isn't too bad, but there are a couple elsewhere in the codebase that *really* threw me.

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (charlespierce) [reply Cargo.toml]: That's reasonable, updated to use `main.rs`
